src/Crypto/Hash/HyperlaneMessageId.hs

```diff
@@ -0,0 +1,161 @@
+
+data TokenMessageERC20 = TokenMessageERC20
+  { tmRecipient :: Text -- variable
+  , tmAmount :: Word256 -- uint256
+  , tmChainId :: Maybe Word256 -- uint256
+  }
+
+packTokenMessageERC20 :: TokenMessageERC20 -> Builder
+packTokenMessageERC20 t =
+  word256BE 64
+  <> word256BE (tmAmount t)
+
+  <> word256BE recipientSize
+  <> BB.byteString recipient
+  where
+    (recipient, recipientSize) = padRight (Text.encodeUtf8 (tmRecipient t))
+
+word256BE :: Word256 -> Builder
+word256BE (Word256 a b c d) =
+  BB.word64BE a <> BB.word64BE b <> BB.word64BE c <> BB.word64BE d
+
+-- | Pad with zeroes on the left to 32 bytes
+--
+-- > padLeft "hello world"
+-- "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NULhello world"
+padLeft :: ByteString -> ByteString
+padLeft s = BS.replicate (32 - BS.length s) 0 <> s
+
+-- | Pad with zeroes on the right, such that the resulting size is a multiple of 32.
+--
+-- > padRight "hello world"
+-- ("hello world\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL",11)
+padRight :: ByteString -> (ByteString, Word256)
+padRight s =
+  let
+    size = BS.length s
+    missingZeroes = restSize size
+  in (s <> BS.replicate missingZeroes 0, fromIntegral size)
+
+-- | Returns the modular of 32 bytes.
+restSize :: Integral a => a -> a
+restSize size = (32 - size) `mod` 32
+
+----------------------------------------------
+--        Hyperlane Message Hashing         --
+----------------------------------------------
+
+getHyperlaneMessageId :: HyperlaneMessage -> Text
+getHyperlaneMessageId =
+  encodeHex
+  . keccak256Hash
+  . BL.toStrict
+  . BB.toLazyByteString
+  . packHyperlaneMessage
+
+keccak256Hash :: ByteString -> ByteString
+keccak256Hash = BSS.fromShort . _getBytesN . _getKeccak256Hash . keccak256
+
+encodeHex :: ByteString -> Text
+encodeHex b = "0x" <> Text.decodeUtf8 (Base16.encode b)
+
+decodeHex :: Text -> Either String ByteString
+decodeHex s
+  | Just h <- Text.stripPrefix "0x" s = Base16.decode (Text.encodeUtf8 h)
+  | otherwise = Left "decodeHex: does not start with 0x"
+
+----------------------------------------------
+--      Hyperlane Pact Object Decoding      --
+----------------------------------------------
+
+decodeHyperlaneMessageObject :: Object Name -> Maybe HyperlaneMessage
+decodeHyperlaneMessageObject o = do
+  let om = _objectMap (_oObject o)
+
+  hmVersion <- om ^? at "version" . _Just . _TLiteral . _1 . _LInteger . to fromIntegral
+  hmNonce <- om ^? at "nonce" . _Just . _TLiteral . _1 . _LInteger . to fromIntegral
+  hmOriginDomain <- om ^? at "originDomain" . _Just . _TLiteral . _1 . _LInteger . to fromIntegral
+  hmSender <- om ^? at "sender" . _Just . _TLiteral . _1 . _LString . to Text.encodeUtf8
+  hmDestinationDomain <- om ^? at "destinationDomain" . _Just . _TLiteral . _1 . _LInteger . to fromIntegral
+  hmRecipient <- om ^? at "recipient" . _Just . _TLiteral . _1 . _LString . to decodeHex . _Right
```

Comment:
`at <x> . _Just . _TLiteral . _1 . <prism>` seems to be a common format here. Why not pull it out?